File: app/src/main/java/com/khch/androidshapes/ShapeFactory.java
package com.khch.androidshapes;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.RectF;

/**
 * Created by kuanghaochuan on 2016/1/14.
 */
public class ShapeFactory {
    public static final int SHAPE_RECT_ROUND_CORNER = 0;
    public static final int SHAPE_CIRCLE = 1;
    public static final int SHAPE_SQUARE = 2;
    private static final int SHAPE_STAR = 3;

    private static ShapeFactory mShapeFactory;

    private Bitmap mCreatedBitmap = null;
    private Bitmap mOriginalBitmap = null;
    private Canvas mCanvas;
    private Paint mPaint;
    private int width;
    private int height;

    private ShapeFactory() {
    }

    public static ShapeFactory getInstance() {
        if (mShapeFactory == null) {
            mShapeFactory = new ShapeFactory();
        }
        return mShapeFactory;
    }

    public Bitmap getBitmap(Bitmap bitmap, int width, int height, int shapeType) {
        init(bitmap, width, height);

        switch (shapeType) {
            case SHAPE_RECT_ROUND_CORNER:
                return createRectRoundCorner();
            case SHAPE_CIRCLE:
                return createCircle();
            case SHAPE_SQUARE:
                return createSquare();
            case SHAPE_STAR:
                return createStar();
            default:
                return bitmap;
        }
    }

    private void init(Bitmap bitmap, int width, int height) {
        this.width = width;
        this.height = height;

        mOriginalBitmap = bitmap;
        mCreatedBitmap = Bitmap.createBitmap(this.width, this.height, Bitmap.Config.ARGB_8888);

        mPaint = new Paint();
        mPaint.setAntiAlias(true);

        mCanvas = new Canvas(mCreatedBitmap);
    }

    private void drawBitmap() {
        mPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        mCanvas.drawBitmap(mOriginalBitmap, null, new RectF(0, 0, width, height), mPaint);
    }

    private Bitmap createCircle() {

        int radius = width < height ? width >> 1 : height >> 1;
        mCanvas.drawCircle(width >> 1, height >> 1, radius, mPaint);

        drawBitmap();

        return mCreatedBitmap;
    }

    private Bitmap createRectRoundCorner() {
        final int corner = 30;
        mCanvas.drawRoundRect(new RectF(0, 0, width, height), corner, corner, mPaint);

        drawBitmap();

        return mCreatedBitmap;
    }

    private Bitmap createSquare() {
        final int length = width < height ? width : height;
        mCanvas.drawRect(new RectF(0, 0, length, length), mPaint);

        drawBitmap();

        return mCreatedBitmap;
    }

    private Bitmap createStar() {
        Path path = new Path();
        path.moveTo(width >> 1, 0);
        path.lineTo(0, height);
        path.lineTo(width, height);
        path.close();

        mCanvas.drawPath(path, mPaint);

        drawBitmap();

        return mCreatedBitmap;
    }

}

File: app/src/main/java/com/khch/androidshapes/ShapeFactory.kt
package com.khch.androidshapes

import android.graphics.*

/**
 * Created by kuanghaochuan on 2016/1/14.
 */
class ShapeFactory private constructor() {

    private var mCreatedBitmap: Bitmap? = null
    private var mOriginalBitmap: Bitmap? = null
    private var mCanvas: Canvas? = null
    private var mPaint: Paint? = null
    private var width: Int = 0
    private var height: Int = 0

    fun getBitmap(bitmap: Bitmap, width: Int, height: Int, shapeType: Int): Bitmap? {
        init(bitmap, width, height)

        when (shapeType) {
            SHAPE_RECT_ROUND_CORNER -> return createRectRoundCorner()
            SHAPE_CIRCLE -> return createCircle()
            SHAPE_SQUARE -> return createSquare()
            SHAPE_STAR -> return createStar()
            else -> return bitmap
        }
    }

    private fun init(bitmap: Bitmap, width: Int, height: Int) {
        this.width = width
        this.height = height

        mOriginalBitmap = bitmap
        mCreatedBitmap = Bitmap.createBitmap(this.width, this.height, Bitmap.Config.ARGB_8888)

        mPaint = Paint()
        mPaint!!.isAntiAlias = true

        mCanvas = Canvas(mCreatedBitmap!!)
    }

    private fun drawBitmap() {
        mPaint!!.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
        mCanvas!!.drawBitmap(mOriginalBitmap!!, null, RectF(0f, 0f, width.toFloat(), height.toFloat()), mPaint)
    }

    private fun createCircle(): Bitmap? {

        val radius = if (width < height) width shr 1 else height shr 1
        mCanvas!!.drawCircle((width shr 1).toFloat(), (height shr 1).toFloat(), radius.toFloat(), mPaint!!)

        drawBitmap()

        return mCreatedBitmap
    }

    private fun createRectRoundCorner(): Bitmap? {
        val corner = 30
        mCanvas!!.drawRoundRect(
            RectF(0f, 0f, width.toFloat(), height.toFloat()),
            corner.toFloat(),
            corner.toFloat(),
            mPaint!!
        )

        drawBitmap()

        return mCreatedBitmap
    }

    private fun createSquare(): Bitmap? {
        val length = if (width < height) width else height
        mCanvas!!.drawRect(RectF(0f, 0f, length.toFloat(), length.toFloat()), mPaint!!)

        drawBitmap()

        return mCreatedBitmap
    }

    private fun createStar(): Bitmap? {
        val path = Path()
        path.moveTo((width shr 1).toFloat(), 0f)
        path.lineTo(0f, height.toFloat())
        path.lineTo(width.toFloat(), height.toFloat())
        path.close()

        mCanvas!!.drawPath(path, mPaint!!)

        drawBitmap()

        return mCreatedBitmap
    }

    companion object {
        val SHAPE_RECT_ROUND_CORNER = 0
        val SHAPE_CIRCLE = 1
        val SHAPE_SQUARE = 2
        private val SHAPE_STAR = 3

        private var mShapeFactory: ShapeFactory? = null

        fun getInstance(): ShapeFactory {
            return mShapeFactory ?: synchronized(this) {
                mShapeFactory ?: ShapeFactory().also {
                    mShapeFactory = it
                }
            }
        }
    }
}
